### PR TITLE
Exo trigger prompt monitoring92 x

### PR DESCRIPTION
**Pull request by gomber**

### CODE REVIEW

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+
+  histname = "photon_phi"; histtitle = "Photon phi";
+  bookME(ibooker,photonPhiME_,histname,histtitle, phi_binning_1.nbins, phi_binning_1.xmin, phi_binning_1.xmax);
+  setTitle(photonPhiME_,"Photon #phi","events / 0.1 rad");
+
+
+  histname = "photon_eta"; histtitle = "Photon eta";
+  bookME(ibooker,photonEtaME_,histname,histtitle, phi_binning_1.nbins, phi_binning_1.xmin, phi_binning_1.xmax);
+  setTitle(photonEtaME_,"Photon #eta","events / 0.1");
+
```
Comment (mtosi) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: aren't you missing the 2D plot ?

File: DQMOffline/Trigger/python/PhotonMonitor_cff.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import FWCore.ParameterSet.Config as cms
+
+from DQMOffline.Trigger.PhotonMonitor_cfi import hltPhotonmonitoring
+
+#HLT_SinglePhoton200_IDTight
+SinglePhoton300_monitoring = hltPhotonmonitoring.clone()
+SinglePhoton300_monitoring.FolderName = cms.string('HLT/Photon/Photon300/')
+SinglePhoton300_monitoring.numGenericTriggerEventPSet.hltPaths = cms.vstring("HLT_Photon300_NoHE_v")
```
Comment (mtosi) [inline DQMOffline/Trigger/python/PhotonMonitor_cff.py]: aren't you apply any offline selection ?

File: DQMOffline/Trigger/python/PhotonMonitor_cff.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+
+# HLT_SinglePhoton175_IDTight
+SinglePhoton175_monitoring = hltPhotonmonitoring.clone()
+SinglePhoton175_monitoring.FolderName = cms.string('HLT/Photon/Photon175/')
+SinglePhoton175_monitoring.numGenericTriggerEventPSet.hltPaths = cms.vstring("HLT_Photon175_v*")
```
Comment (mtosi) [inline DQMOffline/Trigger/python/PhotonMonitor_cff.py]: aren't you apply any offline selection ?

Comment (gomber) [inline DQMOffline/Trigger/python/PhotonMonitor_cff.py]: there is photonselection in the python file photonMonitoring_cfi, which will be created from .cc file.
 desc.add<std::string>("photonSelection", "pt > 145 && eta<1.4442 && hadTowOverEm<0.0597 && full5x5_sigmaIetaIeta()<0.01031 && chargedHadronIso<1.295");

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc (hunk 1 above)
Comment (gomber) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: which 2D plots you want? i am just making trigger efficiency plot as a function of pt, eta and phi, so that should be enough

File: DQMOffline/Trigger/python/PhotonMonitor_cff.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+
+Photon60_monitoring = hltPhotonmonitoring.clone()
+Photon60_monitoring.FolderName = cms.string('HLT/Photon/Photon60/')
+hltPhotonmonitoring.denGenericTriggerEventPSet.hltPaths = cms.vstring()
+hltPhotonmonitoring.denGenericTriggerEventPSet.andOrHlt_ = cms.bool(True)
```
Comment (mtosi) [inline DQMOffline/Trigger/python/PhotonMonitor_cff.py]: I'm confused :(
are you using any HLT paths for the denominator ?
why are you setting
```
hltPhotonmonitoring.denGenericTriggerEventPSet.hltPaths = cms.vstring()
hltPhotonmonitoring.denGenericTriggerEventPSet.andOrHlt_ = cms.bool(True)

```?

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc (hunk 1 above)
Comment (mtosi) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: in addition, are you planning to add variables for the "out-of-time" photon ?

**Comment by mtosi** [comment]
efficiency in the eta-phi plane ;)
  mia

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc (hunk 1 above)
Comment (gomber) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: I mean having eta and phi 1D plot is sufficient. And yes we would like to add out-of-time photons variable for displaced photons , kevin can comment?

**Comment by arunhep** [comment]
i think this PR needs rebase.

**Comment by gomber** [comment]
@arunhep should be fine now..

**Comment by gomber** [comment]
So guys, I am adding 2D plot and will do another commit as Mia suggested and then we would be good to go!

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+  bookME(ibooker,photonVsLS_,histname,histtitle,ls_binning_.nbins, ls_binning_.xmin, ls_binning_.xmax,photon_binning_.xmin, photon_binning_.xmax);
+  setTitle(photonVsLS_,"LS","Photon pT [GeV]");
+
+  histname = "photon_phi"; histtitle = "Photon phi";
+  bookME(ibooker,photonPhiME_,histname,histtitle, phi_binning_1.nbins, phi_binning_1.xmin, phi_binning_1.xmax);
+  setTitle(photonPhiME_,"Photon #phi","events / 0.1 rad");
+
+
+  histname = "photon_eta"; histtitle = "Photon eta";
+  bookME(ibooker,photonEtaME_,histname,histtitle, phi_binning_1.nbins, phi_binning_1.xmin, phi_binning_1.xmax);
```
Comment (mtosi) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: shouldn't be eta_binning ?

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+
+  desc.add<edm::ParameterSetDescription>("numGenericTriggerEventPSet", genericTriggerEventPSet);
+  desc.add<edm::ParameterSetDescription>("denGenericTriggerEventPSet", genericTriggerEventPSet);
+
+  edm::ParameterSetDescription histoPSet;
+  edm::ParameterSetDescription metPSet;
+  fillHistoPSetDescription(metPSet);
+  histoPSet.add<edm::ParameterSetDescription>("photonPSet", metPSet);
+  std::vector<double> bins = {0.,20.,40.,60.,80.,90.,100.,110.,120.,130.,140.,150.,160.,170.,180.,190.,200.,220.,240.,260.,280.,300.,350.,400.,450.,1000.};
+  histoPSet.add<std::vector<double> >("photonBinning", bins);
```
Comment (mtosi) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: I think you should overwrite this in the python config
taking into account the expected turnON

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc (hunk 5 above)
Comment (gomber) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: phi_binning which we are using in the code is very general, and can be used for eta too, its 64 bins between -3.2 to 3.2,  thats why used the same variable

**Comment by mtosi** [comment]
please use the more appropriate for photon obj

thanks
  mia

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc (hunk 5 above)
Comment (gomber) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: Ok I am using the following one now, since we are mainly interested in barrel region for these, so looking at the ECAL tower size, 1.4442/0.087 =  ~17 bins, so I will cover -1.4442 to 1.4442 with 34 bins.

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+  , nelectrons_ ( iConfig.getParameter<int>("nelectrons" ) )
+{
+
+  photonME_.numerator   = nullptr;
+  photonME_.denominator = nullptr;
+  photonME_variableBinning_.numerator   = nullptr;
+  photonME_variableBinning_.denominator = nullptr;
+  photonVsLS_.numerator   = nullptr;
+  photonVsLS_.denominator = nullptr;
+  photonEtaME_.numerator   = nullptr;
```
Comment (mtosi) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: I think the variable binning should be used for eta as well

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+  photonEtaME_.denominator = nullptr;
+  photonPhiME_.numerator   = nullptr;
+  photonPhiME_.denominator = nullptr;
+
```
Comment (mtosi) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: I would suggest to add the 2D plot

File: DQMOffline/Trigger/python/PhotonMonitor_cff.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+# HLT_SinglePhoton175_IDTight
+SinglePhoton175_monitoring = hltPhotonmonitoring.clone()
+SinglePhoton175_monitoring.FolderName = cms.string('HLT/Photon/Photon175/')
+SinglePhoton175_monitoring.numGenericTriggerEventPSet.hltPaths = cms.vstring("HLT_Photon175_v*")
+
+
+Photon60_monitoring = hltPhotonmonitoring.clone()
+Photon60_monitoring.FolderName = cms.string('HLT/Photon/Photon60/')
+Photon60_monitoring.denGenericTriggerEventPSet.hltPaths = cms.vstring()
+Photon60_monitoring.numGenericTriggerEventPSet.hltPaths = cms.vstring("HLT_Photon60_R9Id90_CaloIdL_IsoL_v*")
```
Comment (mtosi) [inline DQMOffline/Trigger/python/PhotonMonitor_cff.py]: are you sure it is available in 2017 ?

File: DQMOffline/Trigger/python/PhotonMonitor_cff.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+Photon60_monitoring.photonSelection = cms.string("pt > 20 && r9() < 0.1 && ((eta<1.4442 && hadTowOverEm<0.0597 && full5x5_sigmaIetaIeta()<0.01031 && chargedHadronIso<1.295) || (eta<2.5 && eta>1.566 && hadTowOverEm<0.0481 && full5x5_sigmaIetaIeta()<0.03013 && chargedHadronIso<1.011))")
+
+Photon60_DisplacedIdL_monitoring = Photon60_monitoring.clone()
+Photon60_DisplacedIdL_monitoring.FolderName = cms.string('HLT/Photon/Photon60_DisplacedIdL/')
+Photon60_DisplacedIdL_monitoring.denGenericTriggerEventPSet.hltPaths = cms.vstring("HLT_Photon60_R9Id90_CaloIdL_IsoL_v*")
+Photon60_DisplacedIdL_monitoring.numGenericTriggerEventPSet.hltPaths = cms.vstring("HLT_Photon60_R9Id90_CaloIdL_IsoL_DisplacedIdL_v*")
+
+Photon60_DisplacedIdL_PFJet350MinPFJet15_monitoring = Photon60_DisplacedIdL_monitoring.clone()
+Photon60_DisplacedIdL_PFJet350MinPFJet15_monitoring.FolderName = cms.string('HLT/Photon/Photon60_DisplacedIdL_PFJet350MinPFJet15/')
+Photon60_DisplacedIdL_PFJet350MinPFJet15_monitoring.denGenericTriggerEventPSet.hltPaths = cms.vstring("HLT_Photon60_R9Id90_CaloIdL_IsoL_v*","HLT_PFHT350MinPFJet15_v*")
```
Comment (mtosi) [inline DQMOffline/Trigger/python/PhotonMonitor_cff.py]: are you sure they are available ?

**Comment by mtosi** [comment]
what about
HLT_Photon50_R9Id90_HE10_IsoM_v8
HLT_Photon75_R9Id90_HE10_IsoM_v8
HLT_Photon90_R9Id90_HE10_IsoM_v8
HLT_Photon120_R9Id90_HE10_IsoM_v8
HLT_Photon165_R9Id90_HE10_IsoM_v9

File: DQMOffline/Trigger/python/PhotonMonitor_cff.py (hunk 9 above)
Comment (gomber) [inline DQMOffline/Trigger/python/PhotonMonitor_cff.py]: yes it is available kevin requested them
https://its.cern.ch/jira/browse/CMSHLT-1327

File: DQMOffline/Trigger/python/PhotonMonitor_cff.py (hunk 10 above)
Comment (gomber) [inline DQMOffline/Trigger/python/PhotonMonitor_cff.py]: see previous message

**Comment by gomber** [comment]
Hi Mia, 

who requested these paths, I didn't see it in Sam mails, may be I missed it?
And I thought the idea of prompt monitoring is for signal paths, so that we can find the issues (if any) fast. Another question, i assume these paths are prescale?

Regards
Bhawna

**Comment by fwyzard** [comment]
We need monitoring for all triggers.

**Comment by gomber** [comment]
Hi Mia, All

I added all the relevant plots as you requested  (2d plots, H/E and r9) and also added other photon paths.
HLT_Photon50_R9Id90_HE10_IsoM_v8
HLT_Photon75_R9Id90_HE10_IsoM_v8
HLT_Photon90_R9Id90_HE10_IsoM_v8
HLT_Photon120_R9Id90_HE10_IsoM_v8
HLT_Photon165_R9Id90_HE10_IsoM_v9

I need to see how to refer to the below PR so that DQMGeneric client producer part i is taken care of :https://github.com/cms-sw/cmssw/pull/18751, but rest is all done.

Regards
Bhawna

**Comment by mtosi** [comment]
great!
i'll give a look at the update as soon as i'll be in the office

for the client, as far as i understood,
it is probably just a change 
cms.EDAnalyzer
-->
cms.EDProducer

@dmitrijus, could you please comment on this ?
thanks !!!

File: Calibration/HcalAlCaRecoProducers/python/ALCARECOHcalCalMinBiasHI_cff.py [hunk 11]
```diff
@@ -79,7 +79,8 @@
 from Configuration.Eras.Modifier_run2_HCAL_2017_cff import run2_HCAL_2017
 run2_HCAL_2017.toModify( hbherecoNoise,
     processQIE11 = cms.bool(True),
-    setNoiseFlagsQIE11 = cms.bool(True),
```
Comment (mtosi) [inline Calibration/HcalAlCaRecoProducers/python/ALCARECOHcalCalMinBiasHI_cff.py]: I think you need to rebase

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+  float* arr = &fbinning[0];
+  me.numerator   = ibooker.book1D(histname+"_numerator",   histtitle+" (numerator)",   nbins, arr);
+  me.denominator = ibooker.book1D(histname+"_denominator", histtitle+" (denominator)", nbins, arr);
+}
+void PhotonMonitor::bookME(DQMStore::IBooker &ibooker, PhotonME& me, const std::string& histname, const std::string& histtitle, int nbinsX, double xmin, double xmax, double ymin, double ymax)
+{
+  me.numerator   = ibooker.bookProfile(histname+"_numerator",   histtitle+" (numerator)",   nbinsX, xmin, xmax, ymin, ymax);
+  me.denominator = ibooker.bookProfile(histname+"_denominator", histtitle+" (denominator)", nbinsX, xmin, xmax, ymin, ymax);
+}
+void PhotonMonitor::bookME(DQMStore::IBooker &ibooker, PhotonME& me, const std::string& histname, const std::string& histtitle, int nbinsX, double xmin, double xmax, int nbinsY, double ymin, double ymax)
```
Comment (dmitrijus) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: These method overloads are very hard to follow, why not just name them for what they are supposed to do?
bookME2D or book2DME or bookProfileME

The main worry of mine is the added memory consumption.

**Review by dmitrijus — CHANGES_REQUESTED** [review]

**Comment by dmitrijus** [comment]
please test

**Comment by mtosi** [comment]
they come from my original example/template, I'm afraid
do they really bring memory consumption named like that ?

**Comment by dmitrijus** [comment]
No. Naming like that just makes it harder to see how many bins are they using.
I will wait for the tests and see it the empirical change, if any :)

**Comment by gomber** [comment]
I tested it again and it worked fine in my area with the little change i committed earlier, so please go ahead and test it.
Sorry about that!

Regards
Bhawna

**Comment by mtosi** [comment]
@dmitrijus could we start the integration test ?

**Comment by gomber** [comment]
Hi Guys?

@dmitrijus Could you please do the integration test, so that we can finalize this PR?

Regards
Bhawna

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+void PhotonMonitor::bookME(DQMStore::IBooker &ibooker, PhotonME& me, const std::string& histname, const std::string& histtitle, int nbins, double min, double max)
+{
+  me.numerator   = ibooker.book1D(histname+"_numerator",   histtitle+" (numerator)",   nbins, min, max);
+  me.denominator = ibooker.book1D(histname+"_denominator", histtitle+" (denominator)", nbins, min, max);
+}
+void PhotonMonitor::bookME(DQMStore::IBooker &ibooker, PhotonME& me, const std::string& histname, const std::string& histtitle, const std::vector<double>& binning)
+{
+  int nbins = binning.size()-1;
+  std::vector<float> fbinning(binning.begin(),binning.end());
+  float* arr = &fbinning[0];
```
Comment (dmitrijus) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: use fbinning.data() instead.

**Review by dmitrijus — CHANGES_REQUESTED** [review]

**Comment by dmitrijus** [comment]
please test

**Comment by fwyzard** [comment]
tracked by #19142

**Comment by mtosi** [comment]
can we proceed w/ this ?

**Comment by mtosi** [comment]
@dmitrijus, if it is fine w/ you, can you please sign it ?
thanks !

**Comment by dmitrijus** [comment]
+1

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc [hunk 14]
```diff
@@ -0,0 +1,350 @@
+#include "DQMOffline/Trigger/plugins/PhotonMonitor.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "DQM/TrackingMonitor/interface/GetLumi.h"
+
+#include "CommonTools/TriggerUtils/interface/GenericTriggerEventFlag.h"
+
+double MAX_PHI1 = 3.2;
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: hi @gomber 
no global variables - just move these into the scope of the Photon Monitor class

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc [hunk 15]
```diff
@@ -0,0 +1,350 @@
...
+  
+}
+
+PhotonMonitor::~PhotonMonitor()
+{
+  if (num_genTriggerEventFlag_) delete num_genTriggerEventFlag_;
+  if (den_genTriggerEventFlag_) delete den_genTriggerEventFlag_;
+}
+
+MEbinning PhotonMonitor::getHistoPSet(edm::ParameterSet pset)
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: pass pset as a const &

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+MEbinning PhotonMonitor::getHistoLSPSet(edm::ParameterSet pset)
+{
+  return MEbinning{
+    pset.getParameter<int32_t>("nbins"),
+      0.,
+      double(pset.getParameter<int32_t>("nbins"))
+      };
+}
+
+void PhotonMonitor::setTitle(PhotonME& me, std::string titleX, std::string titleY)
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: pass the strings as const &

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc (hunk 13 above)
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: it looks like this comment is still to be addressed?

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+  if ( ! metSelection_( pfmet ) ) return;
+  
+  //float met = pfmet.pt();
+  //  float phi = pfmet.phi();
+
+  edm::Handle<reco::PFJetCollection> jetHandle;
+  iEvent.getByToken( jetToken_, jetHandle );
+  std::vector<reco::PFJet> jets;
+  jets.clear();
+  if ( int(jetHandle->size()) < njets_ ) return;
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: make njets_ a unsigned int to avoid the need for int()

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc [hunk 18]
```diff
@@ -0,0 +1,350 @@
...
+  if ( int(jets.size()) < njets_ ) return;
+  
+  edm::Handle<reco::GsfElectronCollection> eleHandle;
+  iEvent.getByToken( eleToken_, eleHandle );
+  std::vector<reco::GsfElectron> electrons;
+  if ( int(eleHandle->size()) < nelectrons_ ) return;
+  for ( auto const & e : *eleHandle ) {
+    if ( eleSelection_( e ) ) electrons.push_back(e);
+  }
+  if ( int(electrons.size()) < nelectrons_ ) return;
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: and similarly here.

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc [hunk 19]
```diff
@@ -0,0 +1,350 @@
...
+	  photonEtaPhiME_.denominator -> Fill(photons[0].eta(), photons[0].phi()); 
+	  photonr9ME_.denominator->Fill(photons[0].r9());
+	  photonHoverEME_.denominator->Fill(photons[0].hadTowOverEm());
+	}
+      
+  // applying selection for numerator
+  if (num_genTriggerEventFlag_->on() && ! num_genTriggerEventFlag_->accept( iEvent, iSetup) ) return;
+
+  // filling histograms (num_genTriggerEventFlag_)  
+  if(photons.size()>0)
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: use photons.empty()

File: DQMOffline/Trigger/plugins/PhotonMonitor.h [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+  PhotonME photonPhiME_;
+  PhotonME photonME_variableBinning_;
+  PhotonME photonVsLS_;
+  PhotonME photonEtaPhiME_;   
+  PhotonME photonr9ME_;   
+  PhotonME photonHoverEME_;   
+
+
+
+  GenericTriggerEventFlag* num_genTriggerEventFlag_;
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/PhotonMonitor.h]: please make these unique_ptrs

File: DQMOffline/Trigger/plugins/PhotonMonitor.h [hunk 21]
```diff
@@ -0,0 +1,121 @@
+#ifndef PhotonMonitor_H
+#define PhotonMonitor_H
+
+#include <string>
+#include <vector>
+#include <map>
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/PhotonMonitor.h]: it looks like this include is not needed

File: DQMOffline/Trigger/plugins/PhotonMonitor.h [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+  PhotonME photonEtaPhiME_;   
+  PhotonME photonr9ME_;   
+  PhotonME photonHoverEME_;   
+
+
+
+  GenericTriggerEventFlag* num_genTriggerEventFlag_;
+  GenericTriggerEventFlag* den_genTriggerEventFlag_;
+
+  StringCutObjectSelector<reco::MET,true>         metSelection_;
```
Comment (davidlange6) [inline DQMOffline/Trigger/plugins/PhotonMonitor.h]: do you expect significantly more complex selections than those in the fillDescriptions? If not, the StringCutObjectSelector is very slow and memory intensive for your needs

Comment (davidlange6) [inline DQMOffline/Trigger/plugins/PhotonMonitor.h]: ah - never mind - I see a complex one now

**Comment by mtosi** [comment]
I think you need a rebase

**Comment by bhawnasinp** [comment]
@davidlange6 you wanted me to use unique_ptr for  GenericTriggerEventFlag* num_genTriggerEventFlag_; I tried using it following way 
std::unique_ptr<GenericTriggerEventFlag> num_genTriggerEventFlag_ (new GenericTriggerEventFlag);  
It gives me error. Do you have any suggestion for the same?

**Comment by davidlange6** [comment]
i don't follow your syntax I guess - you declare num_genTriggerEventFlag_  as a class data member, so you have to change its type in the header, and then change the syntax,similar to below, to set it in the cc. Is that what you did?

**Comment by gomber** [comment]
@davidlange6 so i fixed the issue, and everything is up-to date now from my side. Please have a look!

File: DQMOffline/Trigger/plugins/PhotonMonitor.cc [hunk 23]
```diff
@@ -54,8 +54,8 @@ PhotonMonitor::PhotonMonitor( const edm::ParameterSet& iConfig ) :
 
 PhotonMonitor::~PhotonMonitor()
 {
-  if (num_genTriggerEventFlag_) delete num_genTriggerEventFlag_;
-  if (den_genTriggerEventFlag_) delete den_genTriggerEventFlag_;
+  if (num_genTriggerEventFlag_) num_genTriggerEventFlag_.release();
```
Comment (Dr15Jones) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: This leaks memory. `release()` returns the pointer and stops the ownership`. Instead, just remove lines 57 and 58. In fact, you can do away with the entire destructor since the compiler will do the right thing.

Comment (gomber) [inline DQMOffline/Trigger/plugins/PhotonMonitor.cc]: ok thanks, I will remove the destructor and push it again.

**Comment by dmitrijus** [comment]
+1

**Comment by gomber** [comment]
@cmsbuild Looking at 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18971/20543/unitTests.log
I see, the error is due to
"Input file file:dqm_missing.root appears to be corrupted since it does not have the proper internal structure. Check to see if the file was closed properly."

So I am not sure what can I do for the same?

**Comment by davidlange6** [comment]
the unit test error should not be related to this PR

**Comment by davidlange6** [comment]
hi @gomber - three additional comments (all repeated on N PRs now...)

1) add the package name to the include guard - eg, DQMOFFLINE_TRIGGER_PHOTON_H
2) make nbins an unsigned int everywhere
3) how many new histogram bins does this PR add?

**Comment by gomber** [comment]
@davidlange6  I have implemeted your comments about package name and nbins as using int everywhere. And regarding your 3) point, this PR add 8 new histograms.

**Comment by dmitrijus** [comment]
+1

**Comment by davidlange6** [comment]
hi @gomber - its bins not histograms that we need to check (but to clarify  you have more than 8 histograms.. you have 8 MEs with multiple histograms and ~10 instances of your module. so maybe 160 histograms)

**Comment by gomber** [comment]
Thanks @davidlange6! you are right,s its 160 histograms. Is there somethign more needs to be done or can we merge this PR?

**Comment by davidlange6** [comment]
how many _bins_?

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by gomber** [comment]
@dmitrijus please test again, there was one small bug in the photon_cff file, which i fixed, looks like some typo happened in the last round.

**Comment by gomber** [comment]
@fwyzard here it is https://github.com/cms-sw/cmssw/pull/19506

**Comment by gomber** [comment]
@davidlange6 Regarding the number of bins, we have 16 histograms which corresponds to 2900 bins in total and then there are 10 instances of it, so that would mean in total this code would add 29000 bins, if this is the kind of information you asked me.

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by gomber** [comment]
@davidlange6 Could you please proceed with this? as all the tests are passed and we also provide nbins information?

**Comment by davidlange6** [comment]
all that happened a few hours ago:)

**Comment by davidlange6** [comment]
merge
